discovery/eureka/eureka.go

```diff
@@ -0,0 +1,210 @@
+	RefreshInterval  model.Duration               `yaml:"refresh_interval,omitempty"`
+	HTTPClientConfig config_util.HTTPClientConfig `yaml:",inline"`
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	if len(c.Server) == 0 {
+		return errors.New("eureka_sd: empty or null eureka server")
+	}
+
+	return c.HTTPClientConfig.Validate()
+}
+
+type applicationsClient func(ctx context.Context, server string, client *http.Client) (*Applications, error)
+
+// Discovery provides service discovery based on a Eureka instance.
+type Discovery struct {
+	*refresh.Discovery
+	client      *http.Client
+	server      string
+	lastRefresh map[string]*targetgroup.Group
+	appsClient  applicationsClient
+}
+
+// New creates a new Eureka discovery for the given role.
+func NewDiscovery(conf SDConfig, logger log.Logger) (*Discovery, error) {
+	rt, err := config_util.NewRoundTripperFromConfig(conf.HTTPClientConfig, "eureka_sd", false)
+	if err != nil {
+		return nil, err
+	}
+
+	d := &Discovery{
+		client:     &http.Client{Transport: rt},
+		server:     conf.Server,
+		appsClient: fetchApps,
+	}
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"eureka",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d, nil
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	targetMap, err := d.fetchTargetGroups(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	all := make([]*targetgroup.Group, 0, len(targetMap))
+	for _, tg := range targetMap {
+		all = append(all, tg)
+	}
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	// Remove services which did disappear.
+	for source := range d.lastRefresh {
+		_, ok := targetMap[source]
+		if !ok {
+			all = append(all, &targetgroup.Group{Source: source})
+		}
+	}
+
+	d.lastRefresh = targetMap
+	return all, nil
+}
+
+func (d *Discovery) fetchTargetGroups(ctx context.Context) (map[string]*targetgroup.Group, error) {
+	apps, err := d.appsClient(ctx, d.server, d.client)
+	if err != nil {
+		return nil, err
+	}
+
+	groups := appsToTargetGroups(apps)
+	return groups, nil
+}
+
+// AppsToTargetGroups takes an array of Eureka Applications and converts them into target groups.
+func appsToTargetGroups(apps *Applications) map[string]*targetgroup.Group {
+	tgroups := map[string]*targetgroup.Group{}
+	for _, a := range apps.Applications {
+		group := createTargetGroup(&a)
+		tgroups[group.Source] = group
+	}
+	return tgroups
+}
+
+func createTargetGroup(app *Application) *targetgroup.Group {
+	var (
+		targets = targetsForApp(app)
+		appName = model.LabelValue(app.Name)
+	)
+	tg := &targetgroup.Group{
+		Targets: targets,
+		Labels: model.LabelSet{
+			appNameLabel: appName,
+		},
+		Source: app.Name,
```

Comment:
I don't think that this is the meaning of Source. It should probably just be eureka.

Reply:
I will change source to 'eureka'